presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java

```diff
@@ -224,50 +281,184 @@ public void addInput(Page page)
     {
         checkState(state == State.NEEDS_INPUT, "Operator is already finishing");
         checkNotNull(page, "page is null");
+        checkState(pendingInput == null, "Operator already has pending input");
+
+        if (page.getPositionCount() == 0) {
+            return;
+        }
 
-        pagesIndex.addPage(page);
+        pendingInput = page;
+        if (processPendingInput()) {
+            state = State.HAS_OUTPUT;
+        }
         operatorContext.setMemoryReservation(pagesIndex.getEstimatedSize().toBytes());
     }
 
+    /**
+     * @return true if a full group has been buffered after processing the pendingInput, false otherwise
+     */
+    private boolean processPendingInput()
+    {
+        checkState(pendingInput != null);
+        pendingInput = updatePagesIndex(pendingInput);
+
+        if (pendingInput != null || state == State.FINISHING) {
+            // Seeing unused input or the end of input implies that we have currently buffered a full group
```

Comment:
`If we have unused input or are finishing, then we have buffered a complete partition`